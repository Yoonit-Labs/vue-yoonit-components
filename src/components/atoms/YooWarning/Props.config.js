module.exports = {
  variation: {
    options: [
      'default',
      'clear',
      'danger'
    ]
  },
  iconStyle: {
    options: [
      'solid',
      'regular'
    ]
  }
}
