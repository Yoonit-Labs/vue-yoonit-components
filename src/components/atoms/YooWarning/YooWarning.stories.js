import yooWarning from './YooWarning.vue'
import PropsConfig from './YooWarning.config'

export default {
  title: 'Yoonit/Warning',
  component: yooWarning,
  argTypes: {
    variation: {
      control: {
        type: 'select',
        options: PropsConfig.variation.options
      }
    },
    iconStyle: {
      control: {
        type: 'select',
        options: PropsConfig.iconStyle.options
      }
    }
  }
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { yooWarning },
  template:
    '<yoo-warning v-bind="$props"></yoo-warning>'
})

export const Custom = Template.bind({})
Custom.args = { text: 'Yoo Warning test component' }

export const Default = Template.bind({})
Default.args = { text: 'Yoo Warning test component', variation: 'default' }

export const Clear = Template.bind({})
Clear.args = { text: 'Yoo Warning test component', variation: 'clear' }

export const Danger = Template.bind({})
Danger.args = { text: 'Yoo Warning test component', variation: 'danger' }
