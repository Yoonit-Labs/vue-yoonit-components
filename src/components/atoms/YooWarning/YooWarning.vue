<template lang="pug">
  .warning.p__t--l.p__b--l.m__b--xl(
    :class="[takeModifier]"
  )
    .warning__icon.m__l--l
      i(
        :class="['yoo-warning__icon', takeIconStyle, takeIconName]"
      )

    .warning__body
      p.warning__title.m__t--none.m__b--xs.m__l--l.m__r--l(
        v-show="title"
      ) {{title}}
      p.warning__text.m__t--none.m__b--none.m__l--l.m__r--l {{text}}
</template>

<script>

export default {
  name: 'YooWarning',
  props: {
    title: {
      type: String,
      required: false,
      validator: (value) => {
        return typeof value === 'string'
      },
      default: ''
    },
    text: {
      type: String,
      required: false,
      validator: (value) => {
        return typeof value === 'string'
      },
      default: ''
    },
    icon: {
      type: String,
      required: false,
      validator: (value) => {
        return typeof value === 'string'
      },
      default: 'cog'
    },
    iconStyle: {
      type: String,
      default: 'solid',
      validator (value) {
        return [
          'solid',
          'regular'
        ].indexOf(value) !== -1
      }
    },
    variation: {
      type: String,
      default: 'default',
      validator (value) {
        return [
          'default',
          'clear',
          'danger'
        ].indexOf(value) !== -1
      }
    }
  },
  components: {},
  computed: {
    /**
    * @description Returns formatted icon name
    * @computed takeIconName
    * @returns {string}
    */
    takeIconName () {
      return `fa-${this.icon}`
    },
    /**
    * @description Returns icon style based on iconStyle prop
    * @computed takeIconStyle
    * @returns {string}
    */
    takeIconStyle () {
      return this.iconStyle === 'solid'
        ? 'fas'
        : 'far'
    },
    /**
    * @description Print classes based on the chosen props
    * @computed takeModifier
    * @returns {array}
    */
    takeModifier () {
      const block = 'warning'
      const classList = []

      if (this.variation) {
        classList
          .push(
            `${block}--${this.variation}`
          )
      }

      return classList
    }
  },
  methods: {},
  filters: {},
  watch: {}
}
</script>

<style src="./YooWarning.sass" lang="sass" scoped></style>
