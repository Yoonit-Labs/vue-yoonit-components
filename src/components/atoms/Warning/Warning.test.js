import { shallowMount } from '@vue/test-utils'
import YooWarning from '@/components/atoms/Warning/Warning.vue'
import PropsConfig from './Warning.config'

const classBlock = 'yoo-warning'
const SlotText = 'Default Slot Text'

const mountComponent = () => {
  return shallowMount(YooWarning, {
    slots: { default: SlotText }
  })
}

describe('YooWarning Component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mountComponent()
  })

  it('Matches Snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', () => {
    expect(wrapper.find('.yoo-warning__icon').exists()).toBe(true)
    expect(wrapper.find('.yoo-warning__body').exists()).toBe(true)
  })

  describe('Props', () => {
    describe('variation', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.variation.options.includes(YooWarning.props.variation.default)).toBe(true)
      })
      PropsConfig.variation.options.forEach(variation => {
        it(`Includes variation class: .${classBlock}__variation--${variation}`, async () => {
          await wrapper.setProps({ variation })
          expect(wrapper.find(`.${classBlock}__variation--${variation}`).exists()).toBe(true)
        })
      })
    })

    describe('iconStyle', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.iconStyle.options.includes(YooWarning.props.iconStyle.default)).toBe(true)
      })
      PropsConfig.iconStyle.options.forEach(iconStyle => {
        it(`Includes iconStyle class: .${iconStyle === 'solid' ? 'fas' : 'far'}`, async () => {
          await wrapper.setProps({ icon: 'icon-value', iconStyle })
          expect(wrapper.find('.yoo-warning__icon').classes().includes(`${iconStyle === 'solid' ? 'fas' : 'far'}`)).toBe(true)
        })
      })
    })
  }) // describe Props
})
