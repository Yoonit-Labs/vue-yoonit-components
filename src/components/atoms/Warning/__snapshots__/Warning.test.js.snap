// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooWarning Component Matches Snapshot 1`] = `
<div class="warning p__t--l p__b--l m__b--xl warning--default">
  <div class="warning__icon m__l--l"><i class="yoo-warning__icon fas fa-cog"></i></div>
  <div class="warning__body">
    <p class="warning__title m__t--none m__b--xs m__l--l m__r--l" style="display: none;"></p>
    <p class="warning__text m__t--none m__b--none m__l--l m__r--l"></p>
  </div>
</div>
`;
